### PR TITLE
Fix VS perf issue for create UAP project

### PR DESCRIPTION
**Pull request by jainaashish**
We recently made changes to access all dte calls from UI thread which regressed create UAP project RPS performance. So this PR starts a new task to consume dte through UI thread and doesn't wait for it. It improves existing performance by ~9% but doesn't remove full regressed time. But we tried multiple ways and this is the best we can get given we can't delay these initialization further. If this also doesn't help reach goal for that test then we might need to ask for exception.

Internal bug# 29784

@rrelyea @alpaix @emgarten @rohit21agrawal

### CODE REVIEW

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs [hunk 1]
```diff
@@ -47,21 +50,38 @@ public sealed class RestoreManagerPackage : AsyncPackage
...
+                {
+                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    var dte = (EnvDTE.DTE)await GetServiceAsync(typeof(SDTE));
+                    _buildEvents = dte.Events.BuildEvents;
+                    _buildEvents.OnBuildBegin += BuildEvents_OnBuildBegin;
+
+                    UserAgent.SetUserAgentString(
+                        new UserAgentStringBuilder().WithVisualStudioSKU(dte.GetFullVsVersionString()));
+                }
+                catch (Exception ex)
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs]: Why is this needed? Are you seeing regular crashes from this? I don't see anything that should have a try catch, and logging this to the activity log is pretty much silent.

**Review by emgarten — CHANGES_REQUESTED** [review]
waiting on try/catch question

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs (hunk 1 above)
Comment (jainaashish) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs]: As per my discussion with Alex, this is the recommended guideline when you don't wait for async task to complete. I don't see it failing ever but just following the guideline and also making sure that VS doesn't crash because of this.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs [hunk 2]
```diff
@@ -47,21 +50,38 @@ public sealed class RestoreManagerPackage : AsyncPackage
...
+                {
+                    // log error to activity logs.
+                    ActivityLog.LogError(LogEntrySource, ex.ToString());
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs]:
```c#
ExceptionHelper.WriteToActivityLog(ex);
```

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs (hunk 1 above)
Comment (alpaix) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs]: Any exception in detached task has to be observed.
See [Exception Handling (Task Parallel Library)](https://msdn.microsoft.com/en-us/library/dd997415(v=vs.110).aspx).

Comment (emgarten) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs]: Didn't notice this wasn't blocking at first since there is an await on another initialize below it. I don't think we should launch this without waiting. There is nothing stopping a build from starting before NuGet starts reading it except for pure luck. Potentially 100s of builds could take place before this thread wakes up and hooks up the event. VS automation is like to hit issues with this first.

Comment (jainaashish) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs]: AFAIK build will only start once solution has been loaded, is there a way that a build can start even before solution loads? This thread is hooked up with solution load so i dont see a single case where it will fail. I've also run VS validation build with DDRITs n RPS tests, all have passed.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs (hunk 2 above)
Comment (jainaashish) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs]: updated

Comment (alpaix) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs]: remove the `LogEntrySource` constant as well.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs (hunk 1 above)
Comment (alpaix) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs]: "Thread" is not the right term here. Do you mean "package"?
Also I'm not sure how background load works and whether build can start before all the packages bound to the solution-loaded UI context are initialized.
But with this change new detached and delayed initialization task may complete some time later after the package itself is initialized and solution is loaded. So it does increase the risk we'd miss first build(s).

Comment (jainaashish) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs]: Yes I meant Package...

`But with this change new detached and delayed initialization task may complete some time later after the package itself is initialized and solution is loaded. So it does increase the risk we'd miss first build(s).`
Completely agree there, if we see that way, then it does increase risk of missing first build event.

**Comment by jainaashish** [comment]
@emgarten @alpaix updated

**Comment by emgarten** [comment]
Interesting fix, does this work? doesn't this just put the burden of being on the UI thread on the first caller to these properties, making it all somewhat error prone? Or is the UI thread not needed for GetService?

**Comment by jainaashish** [comment]
@emgarten Yes this works. This was the original code before we changed to normal load. GetService doesn't need to be on UI thread so it doesn't put any additional burden on first caller. Besides these solutionWorker or solutionManager should be load whenever they are needed.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs [hunk 3]
```diff
@@ -34,35 +35,45 @@ public sealed class RestoreManagerPackage : AsyncPackage
         /// </summary>
         public const string PackageGuidString = "2b52ac92-4551-426d-bd34-c6d7d9fdd1c5";
 
-        [Import]
-        private ISolutionRestoreWorker SolutionRestoreWorker { get; set; }
+        private Lazy<ISolutionRestoreWorker> _restoreWorker;
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:38]: Why not 
```c#
[Import]
private Lazy<ISolutionRestoreWorker> SolutionRestoreWorker {get;set;}
```

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs [hunk 4]
```diff
@@ -34,35 +35,45 @@ public sealed class RestoreManagerPackage : AsyncPackage
...
         {
             var componentModel = await GetServiceAsync(typeof(SComponentModel)) as IComponentModel;
             componentModel.DefaultCompositionService.SatisfyImportsOnce(this);
 
-            // Accessing DTE without confirming to Main thread because AllowsBackgroundLoading is false which
-            // will make sure that this piece is always executed on Main thread.
-            var dte = (EnvDTE.DTE)await GetServiceAsync(typeof(SDTE));
-            _buildEvents = dte.Events.BuildEvents;
-            _buildEvents.OnBuildBegin += BuildEvents_OnBuildBegin;
+            _restoreWorker = new Lazy<ISolutionRestoreWorker>(
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:57]: MEF will do this for you when `[Import]` attribute is used  with `Lazy`.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs [hunk 5]
```diff
@@ -34,35 +35,45 @@ public sealed class RestoreManagerPackage : AsyncPackage
...
+                () => componentModel.GetService<ISolutionRestoreWorker>());
+
+            _settings = new Lazy<ISettings>(
+                () => componentModel.GetService<ISettings>());
+
+            _solutionManager = new Lazy<IVsSolutionManager>(
+                () => componentModel.GetService<IVsSolutionManager>());
+
+            await ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:66]: So the task IS awaited now?

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs (hunk 3 above)
Comment (jainaashish) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:38]: I tried that as well, but there was ~10ms difference in perf. Which is why I went ahead with this approach.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs (hunk 4 above)
Comment (jainaashish) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:57]: As replied above, with [Import] Lazy there was a ~10ms perf degradation.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs (hunk 5 above)
Comment (jainaashish) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:66]: Yes

**Review by emgarten — APPROVED** [review]
